contracts/src/Gateway.sol

```diff
@@ -493,199 +391,182 @@ contract Gateway is IGateway, IInitializable, IUpgradable {
+        for (uint256 i = 0; i < message.commands.length; i++) {
+            if (gasleft() * 63 / 64 < message.commands[i].gas + DISPATCH_OVERHEAD_GAS_V2)
+            {
+                assembly {
+                    invalid()
+                }
+            }
+            if (message.commands[i].kind == CommandKind.Upgrade) {
+                try Gateway(this).v2_handleUpgrade{gas: message.commands[i].gas}(
+                    message.commands[i].payload
+                ) {} catch {
+                    return false;
+                }
+            } else if (message.commands[i].kind == CommandKind.SetOperatingMode) {
+                try Gateway(this).v2_handleSetOperatingMode{gas: message.commands[i].gas}(
+                    message.commands[i].payload
+                ) {} catch {
+                    return false;
+                }
+            } else if (message.commands[i].kind == CommandKind.UnlockNativeToken) {
+                try Gateway(this).v2_handleUnlockNativeToken{
+                    gas: message.commands[i].gas
+                }(message.commands[i].payload) {} catch {
+                    return false;
+                }
+            } else if (message.commands[i].kind == CommandKind.RegisterForeignToken) {
+                try Gateway(this).v2_handleRegisterForeignToken{
+                    gas: message.commands[i].gas
+                }(message.commands[i].payload) {} catch {
+                    return false;
+                }
+            } else if (message.commands[i].kind == CommandKind.MintForeignToken) {
+                try Gateway(this).v2_handleMintForeignToken{gas: message.commands[i].gas}(
+                    message.commands[i].payload
+                ) {} catch {
+                    return false;
+                }
+            } else if (message.commands[i].kind == CommandKind.CreateAgent) {
+                try Gateway(this).v2_handleCreateAgent{gas: message.commands[i].gas}(
+                    message.origin
+                ) {} catch {
+                    return false;
+                }
+            } else if (message.commands[i].kind == CommandKind.CallContract) {
+                try Gateway(this).v2_handleCallContract{gas: message.commands[i].gas}(
+                    message.origin, message.commands[i].payload
+                ) {} catch {
+                    return false;
+                }
+            }
         }
+        return true;
     }
 
-    /// @dev Ensure that the specified parachain has a channel allocated
-    function _ensureChannel(ChannelID channelID) internal view returns (Channel storage ch) {
-        ch = CoreStorage.layout().channels[channelID];
-        // A channel always has an agent specified.
-        if (ch.agent == address(0)) {
-            revert ChannelDoesNotExist();
-        }
+    // See docs for `IGateway.sendMessage`
+    function sendMessage(bytes calldata xcm, bytes[] calldata assets) external payable {
```

Comment:
I think we should add a MultiAddress parameter that can be the asset claimer. If provided then we can use this account with the `SetClaimer` XCM instruction. If it is zero/empty address then we SetClaimer to the origin(msg.sender).